pricegraph/src/lib.rs

```diff
@@ -41,3 +44,123 @@ pub struct TransitiveOrder {
     /// The effective sell amount for this transient order.
     pub sell: f64,
 }
+
+impl TransitiveOrder {
+    /// Retrieves the exchange price for this order.
+    pub fn price(&self) -> f64 {
+        self.buy / self.sell
+    }
+
+    /// Retrieves the effective price for this order after fees are condidered.
+    ///
+    /// Note that `effective_price > price`.
+    pub fn effective_price(&self) -> f64 {
+        self.price() * FEE_FACTOR
+    }
+}
+
+/// API entry point for computing price estimates and transitive orderbooks for
+/// a give auction.
+pub struct Pricegraph {
+    full_orderbook: Orderbook,
+    reduced_orderbook: Orderbook,
+}
+
+impl Pricegraph {
+    /// Create a new `Pricegraph` instance given an iterator of auction elements
+    /// for the batch.
+    ///
+    /// The auction elements are in the standard exchange format.
+    pub fn new(elements: impl IntoIterator<Item = Element>) -> Self {
+        let orderbook = Orderbook::from_elements(elements);
+        Pricegraph::from_orderbook(orderbook)
+    }
+
+    /// Create a new `Pricegraph` instance from an `Orderbook`.
+    pub fn from_orderbook(mut orderbook: Orderbook) -> Self {
+        let full_orderbook = orderbook.clone();
+        let reduced_orderbook = {
+            orderbook.reduce_overlapping_orders();
+            orderbook
+        };
+
+        Pricegraph {
+            full_orderbook,
+            reduced_orderbook,
+        }
+    }
+
+    /// Gets a copy of the full orderbook for operations that need to contain
+    /// the existing overlapping transitive orders for accuracy.
+    ///
+    /// This method returns a clone of the reduced orderbook because orderbook
+    /// operations are destructive (as they require filling orders).
```

Comment:
Some copy-pasta-gate here.

Also nanonit, the two sentences could probably be combined.

Reply:
:see_no_evil: